api/PayrollProcessor.Data.Persistence/Features/Employees/EmployeeDetailQueryHandler.cs

```diff
@@ -0,0 +1,84 @@
+using System.Collections.Generic;
+using System.Threading;
+using Ardalis.GuardClauses;
+using LanguageExt;
+using Microsoft.Azure.Cosmos;
+using Newtonsoft.Json.Linq;
+using PayrollProcessor.Core.Domain.Features.Employees;
+using PayrollProcessor.Core.Domain.Intrastructure.Operations.Queries;
+
+using static LanguageExt.Prelude;
+
+namespace PayrollProcessor.Data.Persistence.Features.Employees
+{
+    public class EmployeeDetailQueryHandler : IQueryHandler<EmployeeDetailQuery, EmployeeDetail>
+    {
+        private readonly CosmosClient client;
+
+        public EmployeeDetailQueryHandler(CosmosClient client)
+        {
+            Guard.Against.Null(client, nameof(client));
+
+            this.client = client;
+        }
+
+        public TryOptionAsync<EmployeeDetail> Execute(EmployeeDetailQuery query, CancellationToken token = default)
+        {
+            var iterator = client
+                .GetEmployeesContainer()
+                .GetItemQueryIterator<JObject>(requestOptions: new QueryRequestOptions
+                {
+                    PartitionKey = new PartitionKey(query.EmployeeId.ToString())
+                });
+
+            EmployeeRecord? employeeEntity = null;
+            var payrollEntities = new List<EmployeePayrollRecord>();
+
+            return async () =>
+            {
+                while (iterator.HasMoreResults)
+                {
+                    var response = await iterator.ReadNextAsync(token);
+
+                    foreach (var item in response)
+                    {
+                        string type = item.Value<string>("type") ?? "";
```

Comment:
We store multiple schemes/shapes/types in the same collection.

The 'discriminator' between them is the `type` property, which the `*Record` types populate through the `.Map` methods.

here we want to query for all values matching a partition key, regardless of what `type` they are (both Employees and EmployeePayrolls).

I looked for examples of how to do this but couldn't find any.

The solution here that I came up with was to delay converting the parsed JSON into a POCO until we know the value of `type`.

An alternative would be to `item.ToObject<CosmosDbRecord>()` and read its strongly typed `Type` property, but the call `item.Value<string>("type") ?? ""` does the exact same thing without the cost of an extra `ToObject<>()` call for a value that gets thrown away.